### PR TITLE
ZPLS Echogram Generation README

### PR DESCRIPTION
**Pull request by cwingard**
Updates the README to account for changes to the code.

### CODE REVIEW

File: zpls_echograms/README.md [hunk 1]
```diff
@@ -1,104 +1,193 @@
...
+2020-12-21
+
+**Description**
+
+Utilizes [echopype](https://echopype.readthedocs.io/en/latest/), the open-source ocean sonar data processing module, to
+process OOI bioacoustic sonar data from both uncabled and cabled array sensors and generate weekly echograms that 
+the user community can use to quickly assess the quality and applicability of the data to their research needs.
+
+OOI utilizes the [ASL Environmental Sciences Acoustic Zooplankton Fish Profiler (AZFP)](https://aslenv.com/azfp.html) 
+for most uncabled sites (e.g. [CP03ISSM](https://oceanobservatories.org/site/cp03issm/) or
```
Comment (spetillo) [inline zpls_echograms/README.md]: e.g.**,**

File: zpls_echograms/README.md [hunk 2]
```diff
@@ -1,104 +1,193 @@
...
+  and the other looking down).
+
+* The preset colorbar ranges are the same for every array. Feel free to change it via the optional colorbar_range input.
+  
+**Raw and Processed Data Folders and Files:**
+
+The raw ZPLS data can be found on the [OOI Raw Data server](https://rawdata.oceanobservatories.org/files/), with the 
+directory structure dependent on the ZPLS instrument type (either an EK60 or AZFP).
+
+For the AZFP, the data is organized by the instrument in a folder called `DATA/`, below which the files are organized
```
Comment (spetillo) [inline zpls_echograms/README.md]: data **are** organized

File: zpls_echograms/README.md [hunk 3]
```diff
@@ -1,104 +1,193 @@
...
+
+**Notes:** 
+
+* The root data directory depends on the data source. If looking at cabled data and the EK60, the root data directory 
+  is the year (e.g. `2017/`). For uncabled data, the root data directory is `DATA/`.
+ 
+* The path to the `.xml` file is required if the ZPLS model is an AZFP. The `.xml` file contains instrument specific
+  settings and calibration coefficients needed to process the data.
+
+* In the case of the global moorings, the OOI 8-letter site name is appended with either an "_Upper" or "_Lower" to
```
Comment (spetillo) [inline zpls_echograms/README.md]: In the case of the global **Apex Profiler (subsurface)** moorings

File: zpls_echograms/README.md [hunk 4]
```diff
@@ -1,104 +1,193 @@
...
+* The preset colorbar ranges are the same for every array. Feel free to change it via the optional colorbar_range input.
+  
+**Raw and Processed Data Folders and Files:**
+
+The raw ZPLS data can be found on the [OOI Raw Data server](https://rawdata.oceanobservatories.org/files/), with the 
+directory structure dependent on the ZPLS instrument type (either an EK60 or AZFP).
+
+For the AZFP, the data is organized by the instrument in a folder called `DATA/`, below which the files are organized
+in folders named according the year and month (e.g. `201610`). The files are named according to the year, month, day
+and hour the data was recorded (e.g. `16101401.01A`). The higher level directory structure above `DATA/` is created by
```
Comment (spetillo) [inline zpls_echograms/README.md]: the data **were** recorded

File: zpls_echograms/README.md [hunk 5]
```diff
@@ -1,104 +1,193 @@
...
+**Raw and Processed Data Folders and Files:**
+
+The raw ZPLS data can be found on the [OOI Raw Data server](https://rawdata.oceanobservatories.org/files/), with the 
+directory structure dependent on the ZPLS instrument type (either an EK60 or AZFP).
+
+For the AZFP, the data is organized by the instrument in a folder called `DATA/`, below which the files are organized
+in folders named according the year and month (e.g. `201610`). The files are named according to the year, month, day
+and hour the data was recorded (e.g. `16101401.01A`). The higher level directory structure above `DATA/` is created by 
+OOI technicians. If the ZPLS data from the [Global Argentine Basin, upward looking sensor for Deployment 
+3](https://rawdata.oceanobservatories.org/files/GA02HYPM/R00003/instruments/ZPLSG_sn55067/) was downloaded to a local
```
Comment (spetillo) [inline zpls_echograms/README.md]: **were** downloaded to

File: zpls_echograms/README.md [hunk 6]
```diff
@@ -1,104 +1,193 @@
...
+./GA/ZPLSG_sn55067/DATA/201611
+  ./GA/ZPLSG_sn55067/DATA/201611/16111401.01A
+  ./GA/ZPLSG_sn55067/DATA/201611/16111402.01A
+  ./GA/ZPLSG_sn55067/DATA/201611/16111403.01A
+  ./GA/ZPLSG_sn55067/DATA/201611/*.01A
 ```
 
-**Example 1:**  
-```python3    zpls_echogram.py    ./GI/ZPLSG_sn55067/DATA    ./GI/ZPLSG_sn55067/DATA/201610/16101417.XML    20161115    20161116    ./processed_folder```
+For the EK60, the data is organized by OOI technicians in folders based on the year, month and day (e.g. `2017/01/01`).
```
Comment (spetillo) [inline zpls_echograms/README.md]: data **are** organized

File: zpls_echograms/README.md [hunk 7]
```diff
@@ -1,104 +1,193 @@
...
+Files are named by the instrument according to the date and time the file was created (e.g. 
+`OOI-D20170831-T113333.raw`). If the ZPLS data from the [Coastal Endurance, Oregon Offshore Cabled Shallow Profiler 
+Mooring, 200 m Platform](https://rawdata.oceanobservatories.org/files/CE04OSPS/PC01B/ZPLSCB102_10.33.10.143) 
+from 2017 was downloaded to local folder named `CE04`, the folders would look like:
```
Comment (spetillo) [inline zpls_echograms/README.md]: from 2017 **were** downloaded

File: zpls_echograms/README.md [hunk 8]
```diff
@@ -1,104 +1,193 @@
...
+  ./CE04/PC01B/ZPLSCB102_10.33.10.143/2017/01/02/OOI-D20170102-T000000.bot
+  ./CE04/PC01B/ZPLSCB102_10.33.10.143/2017/01/02/OOI-D20170102-T000000.idx
+  ./CE04/PC01B/ZPLSCB102_10.33.10.143/2017/01/02/OOI-D20170102-T000000.raw
+```
 
-The code will process the data from 11/15/2016 to 11/16/2016. The .nc file and .png plot will be stored at ./processed_folder.
+For the AZFP example above, the root data directory would be `./GA/ZPLSG_sn55067/DATA`. For the EK60 example, the root 
+data directory would be `./CE04/PC01B/ZPLSCB102_10.33.10.143/2017`. 
+
+Processed data is created in a folder of the users choice. We recommend placing the data at the same level as the root
```
Comment (spetillo) [inline zpls_echograms/README.md]: data **are**

File: zpls_echograms/README.md [hunk 9]
```diff
@@ -1,104 +1,193 @@
...
+  ./CE04/PC01B/ZPLSCB102_10.33.10.143/2017/01/02/OOI-D20170102-T000000.idx
+  ./CE04/PC01B/ZPLSCB102_10.33.10.143/2017/01/02/OOI-D20170102-T000000.raw
+```
 
-The code will process the data from 11/15/2016 to 11/16/2016. The .nc file and .png plot will be stored at ./processed_folder.
+For the AZFP example above, the root data directory would be `./GA/ZPLSG_sn55067/DATA`. For the EK60 example, the root 
+data directory would be `./CE04/PC01B/ZPLSCB102_10.33.10.143/2017`. 
+
+Processed data is created in a folder of the users choice. We recommend placing the data at the same level as the root 
+data directory in a separate folder called `processed`. The data in the `processed` folder is organized into subfolders
```
Comment (spetillo) [inline zpls_echograms/README.md]: data in the `processed` folder **are** organized

File: zpls_echograms/README.md [hunk 10]
```diff
@@ -1,104 +1,193 @@
...
+There are three types of processed data created by the zpls_echogram script:
+
+1. **Converted**: These are files converted from the raw format (either `.01A` or `.raw`) to NetCDF format (`.nc`) by 
+   the echopype convert function. They are 1:1 with the source raw data file. These `.nc` files use the 
+   [SONAR-NetCDF4](http://www.ices.dk/sites/pub/Publication%20Reports/Cooperative%20Research%20Report%20(CRR)/CRR341.pdf)
+   formatting convention.
+2. **Processed**: The echopype process function, taking the converted files as inputs, is used to calculate the volume 
+   acoustic backscatter strength (Sv), and the range (distance from the sensor face) in meters. This subset of data is 
+   saved in daily `.nc` files that use the site and date range to construct the file name 
+   (e.g. data for 2016-01-07 would be saved in
```
Comment (spetillo) [inline zpls_echograms/README.md]: e.g.**,**

File: zpls_echograms/README.md [hunk 11]
```diff
@@ -1,104 +1,193 @@
...
+   formatting convention.
+2. **Processed**: The echopype process function, taking the converted files as inputs, is used to calculate the volume 
+   acoustic backscatter strength (Sv), and the range (distance from the sensor face) in meters. This subset of data is 
+   saved in daily `.nc` files that use the site and date range to construct the file name 
+   (e.g. data for 2016-01-07 would be saved in 
+   `GA02HYPM_Upper_Bioacoustic_Echogram_20170101-20170108_Calibrated_Sv_Full_20170107.nc`).
+3. **Averaged**: Takes the processed data one step further and applies temporal averaging (15 minute median averages for
+   the coastal ZPLSC and 60 minute median averages for the global ZPLSG) for use in creating the echograms. Missing data
+   is filled in via linear interpolation for gaps less than 3x the averaging time window. The averaged data is saved 
+   in a `.nc` file using a similar file naming convention as the processed data above (e.g.
```
Comment (spetillo) [inline zpls_echograms/README.md]: e.g.**,**

File: zpls_echograms/README.md [hunk 12]
```diff
@@ -1,104 +1,193 @@
...
+   (e.g. data for 2016-01-07 would be saved in 
+   `GA02HYPM_Upper_Bioacoustic_Echogram_20170101-20170108_Calibrated_Sv_Full_20170107.nc`).
+3. **Averaged**: Takes the processed data one step further and applies temporal averaging (15 minute median averages for
+   the coastal ZPLSC and 60 minute median averages for the global ZPLSG) for use in creating the echograms. Missing data
+   is filled in via linear interpolation for gaps less than 3x the averaging time window. The averaged data is saved 
+   in a `.nc` file using a similar file naming convention as the processed data above (e.g. 
+   `GA02HYPM_Upper_Bioacoustic_Echogram_20170101-20170108_Calibrated_Sv_Averaged.nc`).
+   
+An echogram covering the time period specified by the date range is created from the averaged data and saved along with
+the three data types list above. The files share the same naming convention as the processed and averaged data (e.g.
```
Comment (spetillo) [inline zpls_echograms/README.md]: list --> listed
e.g.**,**

File: zpls_echograms/README.md [hunk 13]
```diff
@@ -1,104 +1,193 @@
...
+
+**Example 1:**
+```
+# set default data directories and .xml file 
+GA_DATA="./GA/ZPLSG_sn55067/DATA"
+GA_PROC="./GA/ZPLSG_sn55067/processed"
+GA_XML="./GA/ZPLSG_sn55067/DATA/201610/16101417.XML"
+
+# set the site, the script will look at the mooring configuration and apply the 
+# tilt angle correction (15 degrees for GA02HYPM_Upper) and the theoretical
```
Comment (snwhite) [inline zpls_echograms/README.md]: Perhaps "design depth" would be better than "theoretical depth"

**Review by snwhite — COMMENTED** [review]
Looks pretty good to me.  I had just one wording comment.

File: zpls_echograms/README.md [hunk 14]
```diff
@@ -1,104 +1,193 @@
...
-
-The optional flag -cr will set the colorbar range from -120 to -20.
-
-**Output Naming Convention:**  
-The naming convention is:
+```python zpls_echogram.py -s $SITE -d $GA_DATA -o $GA_PROC -xf $GA_XML -dr 201611 -zm AZFP -dd 160```
 
-subsite + deployment_number + serial_number + date
+If the instrument is deployed at 160 m, instead of the theoretical depth of 150 m. The optional flag `-dd` 160 will 
+overwrite the mooring configuration.
```
Comment (spetillo) [inline zpls_echograms/README.md]: Suggest:
The optional flag **`-dd 160`** will 
**override** the mooring configuration.

**Review by spetillo — CHANGES_REQUESTED** [review]
@cwingard  I like the updates to this. The grammar police stopped by for a visit, and I made a few other minor suggestions, but otherwise this is looking great.

File: zpls_echograms/README.md (hunk 13 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Agreed, made the change here and further down in Example 4.

File: zpls_echograms/README.md (hunk 1 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 2 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 3 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 4 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 5 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 6 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 7 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 8 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 9 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 10 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 11 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 12 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

File: zpls_echograms/README.md (hunk 14 above)
Comment (cwingard) [reply zpls_echograms/README.md]: Changes made

**Review by spetillo — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**